README.md
# Skript na testovanie example web aplikácie s použitím selenium a behave (školský projekt)


features/changes.txt
content_creations.feature:
    Deleted:
        copying
        pasting
    Changed:
        Bad input removed first Then
        is admin always first

references.feature:
    Deleted:
        Input non existing tool reference to a method

visiblity.feature:
    Deleted:
        Non published content item is not visible to a consumer

features/environment.py
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
from selenium.webdriver.common.by import By
from selenium.common.exceptions import NoSuchElementException
from selenium.common.exceptions import NoAlertPresentException
from selenium.common.exceptions import WebDriverException

# ...

URL = "http://localhost:8080/VALU3S"


def ensure_driver():
    '''Get Chrome/Firefox driver from Selenium Hub'''
    try:
        driver = webdriver.Remote(
                command_executor='http://localhost:4444/wd/hub',
                desired_capabilities=DesiredCapabilities.CHROME)
    except WebDriverException:
        driver = webdriver.Remote(
                command_executor='http://localhost:4444/wd/hub',
                desired_capabilities=DesiredCapabilities.FIREFOX)
    driver.implicitly_wait(5)
    return driver

def before_all(context):
    context.driver = ensure_driver()
    context.base_url = URL

def after_all(context):
    context.driver.quit()


features/steps/references.py
from behave import *
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains 
from selenium.webdriver.common.keys import Keys
import time

def logout(context):
    
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.2)
    context.driver.find_element(By.XPATH, "//ul[2]/li/a/span[2]").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "personaltools-logout").click()


@given(u'user is on the method edit page.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1/edit")

@when(u'user selects Relations tab.')
def step_impl(context):
    context.driver.find_element(By.ID, "autotoc-item-autotoc-2").click()


@then(u'user is on the relation modification form.')
def step_impl(context):
    elements = context.driver.find_elements(By.CSS_SELECTOR, "#formfield-form-widgets-related_methods > .horizontal")
    assert len(elements) > 0
    logout(context)


@given(u'some method already exists.')
def step_impl(context):
    pass


@given(u'user is on method relation modification form.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1/edit#autotoc-item-autotoc-2")


@when(u'user inputs name of his desired tool.')
def step_impl(context):
    context.driver.find_element(By.CSS_SELECTOR, "#formfield-form-widgets-tools .path-wrapper .glyphicon").click()
    context.driver.find_element(By.ID, "s2id_autogen12").send_keys("Combine")
    context.driver.find_element(By.CSS_SELECTOR, ".pattern-relateditems-result-title").click()
    context.driver.find_element(By.ID, "form-buttons-save").click()


@when(u'user selects it and click Save.')
def step_impl(context):
    pass


@then(u'related tool has been added to the method.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1/edit#autotoc-item-autotoc-2")
    elements = context.driver.find_elements(By.CSS_SELECTOR, ".pattern-relateditems-item-title")
    assert len(elements) > 0
    logout(context)


@given(u'is at edit method\'s relations form.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1/edit#autotoc-item-autotoc-2")


@when(u'user clicks on small x in the top left corner of tool and clicks Save.')
def step_impl(context):
    context.driver.find_element(By.CSS_SELECTOR, ".select2-search-choice-close").click()
    context.driver.find_element(By.ID, "form-buttons-save").click()


@then(u'reference to that tool has been removed.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1/edit#autotoc-item-autotoc-2")
    elements = context.driver.find_elements(By.CSS_SELECTOR, ".pattern-relateditems-item-title")
    assert len(elements) == 0
    logout(context)


@given(u'user is consumer.')
def step_impl(context):
    pass


@given(u'method is created and published.')
def step_impl(context):
    pass


@given(u'published method references a tool.')
def step_impl(context):
    pass


@when(u'user clicks to list of methods.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S")
    context.driver.find_element(By.LINK_TEXT, "Methods").click()


@when(u'selects said method.')
def step_impl(context):
    context.driver.find_element(By.LINK_TEXT, "Combinatorial Testing").click()


@when(u'clicks on referenced too.')
def step_impl(context):
    context.driver.find_element(By.CSS_SELECTOR, ".contenttype-tool").click()


@then(u'page of that tool will open.')
def step_impl(context):
    assert context.driver.find_element(By.CSS_SELECTOR, ".documentFirstHeading").text == "Testos / Combine"

features/steps/content_creations.py
from behave import *
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains 
from selenium.webdriver.common.keys import Keys
import time

def logout(context):
    
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.2)
    context.driver.find_element(By.XPATH, "//ul[2]/li/a/span[2]").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "personaltools-logout").click()


@given(u'user is an administrator.')
def step_impl(context):
    
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.3)
    context.driver.find_element(By.ID, "personaltools-login").click()
    time.sleep(0.3)
    context.driver.find_element(By.ID, "__ac_name").click()
    context.driver.find_element(By.ID, "__ac_name").send_keys("itsadmin")
    context.driver.find_element(By.ID, "__ac_password").click()
    context.driver.find_element(By.ID, "__ac_password").send_keys("itsadmin")
    time.sleep(0.3)
    context.driver.find_element(By.CSS_SELECTOR, ".pattern-modal-buttons > #buttons-login").click()


@given(u'user is at methods page.')
def step_impl(context):
    pass


@when(u'user clicks "Add new" in side panel and selects method.')
def step_impl(context):
  
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.2)
    context.driver.find_element(By.XPATH, "//li[4]/a/span[2]").click()
    time.sleep(0.2)
    context.driver.find_element(By.LINK_TEXT, "Method").click()
  

@then(u'Method creation form shows up.')
def step_impl(context):
 
    assert context.driver.find_element(By.CSS_SELECTOR, ".documentFirstHeading").text == "Add Method"
    logout(context)


@given(u'is on method creation form.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/++add++method")


@when(u'user fills out form with correct information and clicks "Save".')
def step_impl(context):
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").click()
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").send_keys("Test Method")
    context.driver.find_element(By.ID, "form-widgets-method_purpose").click()
    context.driver.find_element(By.ID, "form-widgets-method_purpose").send_keys("Test method")
    context.driver.switch_to.frame(3)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.switch_to.frame(2)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.switch_to.frame(1)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    context.driver.switch_to.default_content()
    context.driver.execute_script("window.scrollTo(0,2550)")
    context.driver.switch_to.frame(1)
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.find_element(By.ID, "form-buttons-save").click()


@then(u'method is created.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")
    time.sleep(0.3)
    elements = context.driver.find_elements(By.LINK_TEXT, "Test Method")
    assert len(elements) > 0
    logout(context)


@given(u'at least one method exists.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/++add++method")
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").click()
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").send_keys("Test Method")
    context.driver.find_element(By.ID, "form-widgets-method_purpose").click()
    context.driver.find_element(By.ID, "form-widgets-method_purpose").send_keys("Test method")
    context.driver.switch_to.frame(3)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.switch_to.frame(2)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.switch_to.frame(1)
    context.driver.find_element(By.CSS_SELECTOR, "html").click()
    context.driver.switch_to.default_content()
    context.driver.execute_script("window.scrollTo(0,2550)")
    context.driver.switch_to.frame(1)
    element = context.driver.find_element(By.ID, "tinymce")
    context.driver.execute_script("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '<p>test method<br data-mce-bogus=\"1\"></p>'}", element)
    context.driver.switch_to.default_content()
    context.driver.find_element(By.ID, "form-buttons-save").click()


@given(u'user is on page of specific method.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method")
    


@when(u'user clicks on "Edit" button in the sidebar.')
def step_impl(context):
    context.driver.find_element(By.CSS_SELECTOR, "#contentview-edit span:nth-child(2)").click()
    


@then(u'edit method form will show up.')
def step_impl(context):
    assert context.driver.find_element(By.CSS_SELECTOR, ".documentFirstHeading").text == "Edit Method"
    logout(context)


@given(u'user is on the method edit form.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method/edit")


@when(u'user changes one or more fields and cliks "Save".')
def step_impl(context):
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").clear()
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").click()
    context.driver.find_element(By.ID, "form-widgets-IBasic-title").send_keys("edited")
    context.driver.find_element(By.ID, "form-buttons-save").click()


@then(u'Method has been modified.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")
    elements = context.driver.find_elements(By.LINK_TEXT, "edited")
    assert len(elements) > 0
    logout(context)


@given(u'user is on home page.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/")


@when(u'user selects "Contents" from the side bar.')
def step_impl(context):
    context.driver.find_element(By.XPATH, "//li[@id='contentview-folderContents']/a/span").click()


@then(u'user is on content page.')
def step_impl(context):
    assert("http://localhost:8080/VALU3S/folder_contents" in context.driver.current_url)
    logout(context)


@given(u'user is on content page.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")


@given(u'user in in a folder where specific item is located.')
def step_impl(context):
    pass


@when(u'user clicks on checkbox next to method and clicks on delete icon and confirms.')
def step_impl(context):
    context.driver.find_element(By.ID, "select6InputCheckbox").click()
    context.driver.find_element(By.CSS_SELECTOR, ".glyphicon-trash").click()
    context.driver.find_element(By.CSS_SELECTOR, ".applyBtn").click()


@then(u'item is deleted.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")
    elements = context.driver.find_elements(By.LINK_TEXT, "edited")
    assert len(elements) == 0
    logout(context)



@given(u'user is on the method creation from.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.2)
    context.driver.find_element(By.XPATH, "//li[4]/a/span[2]").click()
    time.sleep(0.2)
    context.driver.find_element(By.LINK_TEXT, "Method").click()


@when(u'user don\'t fill in the title of a method and saves.')
def step_impl(context):
    context.driver.find_element(By.ID, "form-buttons-save").click()
    


@then(u'Prompt shows up.')
def step_impl(context):
    elements = context.driver.find_elements(By.CSS_SELECTOR, ".portalMessage")
    assert len(elements) > 0
    context.driver.find_element(By.CSS_SELECTOR, ".portalMessage").click()
    assert context.driver.find_element(By.CSS_SELECTOR, "dd").text == "There were some errors."
    logout(context)


@given(u'a method has been created.')
def step_impl(context):
    pass


@when(u'user goes to a methods page.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1")


@then(u'created method shoud be visible as unpublished (red).')
def step_impl(context):
   assert context.driver.find_element(By.CSS_SELECTOR, ".plone-toolbar-state-title").text == "Private"
   logout(context)




features/steps/visiblity.py
from behave import *
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains 
from selenium.webdriver.common.keys import Keys
import time

def logout(context):
    
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.2)
    context.driver.find_element(By.XPATH, "//ul[2]/li/a/span[2]").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "personaltools-logout").click()


@given(u'unpublished content item is created.')
def step_impl(context):
    pass


@when(u'user selects "Submit for publication".')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1")
    time.sleep(0.2)
    context.driver.find_element(By.CSS_SELECTOR, ".label-state-private > span:nth-child(2)").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "workflow-transition-submit").click()
  


@then(u'content item will change it\'s review state to pending.')
def step_impl(context):
    assert context.driver.find_element(By.CSS_SELECTOR, ".plone-toolbar-state-title").text == "Pending review"
    logout(context)


@given(u'user is a reviewer.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S")
    time.sleep(0.3)
    context.driver.find_element(By.ID, "personaltools-login").click()
    time.sleep(0.3)
    context.driver.find_element(By.ID, "__ac_name").click()
    context.driver.find_element(By.ID, "__ac_name").send_keys("itsreviewer")
    context.driver.find_element(By.ID, "__ac_password").click()
    context.driver.find_element(By.ID, "__ac_password").send_keys("itsreviewer")
    time.sleep(0.3)
    context.driver.find_element(By.CSS_SELECTOR, ".pattern-modal-buttons > #buttons-login").click()


@given(u'content item with pending review status has been created.')
def step_impl(context):
    pass


@when(u'user opens list of content items.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")


@then(u'user will see that content item (green color).')
def step_impl(context):
    elements = context.driver.find_elements(By.LINK_TEXT, "Test Method")
    assert len(elements) > 0
    assert context.driver.find_element(By.LINK_TEXT, "Test Method").text == "Test Method"
    assert context.driver.find_element(By.CSS_SELECTOR, ".state-pending > .review_state").text == "pending"
    logout(context)


@given(u'content item has been created.')
def step_impl(context):
    pass


@given(u'that content item has pending rewiew status.')
def step_impl(context):
    pass


@given(u'user has that content item opened.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1")


@when(u'he selects "Publish" option from sidebar menu.')
def step_impl(context):
    time.sleep(0.2)
    context.driver.find_element(By.CSS_SELECTOR, ".plone-toolbar-state-title").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "workflow-transition-publish").click()


@then(u'content item will change it\'s state to published.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")
    elements = context.driver.find_elements(By.LINK_TEXT, "Test Method")
    assert len(elements) > 0
    assert context.driver.find_element(By.LINK_TEXT, "Test Method").text == "Test Method"
    assert context.driver.find_element(By.CSS_SELECTOR, ".type-method > .review_state").text == "published"
    logout(context)


@given(u'content item was created.')
def step_impl(context):
    pass


@given(u'it was not published or sent to review.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/test-method-1")
    time.sleep(0.2)
    context.driver.find_element(By.CSS_SELECTOR, ".plone-toolbar-state-title").click()
    time.sleep(0.2)
    context.driver.find_element(By.ID, "workflow-transition-reject").click()


@when(u'user open list of content items of a certain type.')
def step_impl(context):
    context.driver.get("http://localhost:8080/VALU3S/folder_contents")


@then(u'user can\'t see that content item.')
def step_impl(context):
    elements = context.driver.find_elements(By.LINK_TEXT, "Test Method")
    assert len(elements) == 0